### PR TITLE
security feature

### PR DESCRIPTION
**Pull request by DanmarGreen0**
Note: see this link [https://github.com/DanmarGreen0/online-shopping-rest-api/blob/security/postman/security-test.postman_collection.json] for my postman tests for my application.

## Risk and Impact Analysis
Users need a way of securing their accounts from other users. Without a security functionality, different users may potentially steal other user accounts or information. Therefore, security is a vital feature of any computer application, especially web applications that store important or confidential data. Also, know would ever want to use an application that exposes their information to the entire world.

## Overview
I have implemented a security feature that allows users to log in to their accounts. I have also implemented web service endpoint security that enforces user access. However, once a user logs into their account using a web browser, they can't log out of their account unless they end the browser session. In a use case where multiple users interact with a single device to sign in to their accounts, this security flaw would leave any user at risk of another user using their account. 

## New feature/API (non-breaking change which adds functionality)
- security
  - user login
  - admin user can view all user accounts and their information
  - basic user can only view their account information

## Reviewer Notes
Checklist:

### CODE REVIEW

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/AuthGroup.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module;
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/AuthGroup.java]: Consider a more concise package structure for readability / usability. E.g. com.onlinesellingandshopping.model or com.onlinesellingandshopping.data . Something that can help another developer get a mental picture of the project structure quickly. Rule of thumb I use is removing any redundant words.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/AuthGroup.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import javax.persistence.*;
+
+@Entity
+@Table(name="Auth_User_Group")
+public class AuthGroup {
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/AuthGroup.java]: Consider using a Builder pattern, take a look here https://stackoverflow.com/questions/8111461/how-can-you-use-builder-pattern-for-entities-with-jpa. You can find the best practice reference in Effective Java Third Edition Item 2. 
This approach makes the code that creates an Entry of this type less prone for errors, more readable, and also more pleasant to write :-)

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/AuthGroup.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name="Auth_User_Group_Id")
+    private Integer id;
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/AuthGroup.java]: Generally speaking, it is best to make fields final, and initialize via a private constructor used by the Builder class. I am not sure if there is a limitation of the JPA library that requires non final fields.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/User.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
+package com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module;
+
+
+
+import javax.persistence.*;
+import java.sql.Date;
+
+@Entity
+public class User {
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/module/User.java]: Feedback provided for AuthGroup applies here as well.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package com.project.portfolio.springboot.onlinesellingandshopping.serverapp.repository;
+
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.Optional;
+
+@Repository
+public interface UserRepository extends JpaRepository<User,Integer> {
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/repository/UserRepository.java]: Missing Javadoc for the method findByUsername. See Effective Java Third Edition Item 56.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.util.List;
+
+@Service
+public class ApplicationUserDetailsService implements UserDetailsService {
+
+    private final UserRepository userRepository;
+    private final AuthGroupRepository authGroupRepository;
+
+    public ApplicationUserDetailsService(UserRepository userRepository, AuthGroupRepository authGroupRepository) {
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java]: Missing null checks.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        this.userRepository = userRepository;
+        this.authGroupRepository = authGroupRepository;
+    }
+
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java]: Missing null check for the username argument.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        User user = (User) this.userRepository.findByUsername(username).get();
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java]: Assuming the return value is of type Optional<User> ? In this case you want to check it's not empty first before retrieving the value.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+        if(null==user){
+            throw new UsernameNotFoundException("couldn't find username: " + username);
+        }
+
+        List<AuthGroup> authGroups = this.authGroupRepository.findByUsername(username);
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserDetailsService.java]: You can drop 'this' , any class member can be used in the class methods without explicitly using 'this' reference. ( unless there is a local variable with the same name ). Also it's best to add final when a local variable is defined and assigned value at the same time ( unless you intend to re-assign the value again later in the method block )

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.*;
+
+public class ApplicationUserPrincipal implements UserDetails {
+
+    private final User user;
+    private final List<AuthGroup> authGroups;
+
+    public ApplicationUserPrincipal(User user, List<AuthGroup> authGroups) {
+        super();
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java]: what is the purpose of super() here?

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        this.user = user;
+        this.authGroups = authGroups;
+    }
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java]: If I understand this logic correctly, it looks like grantedAuthorities depend only on authGroups, in this case you can make this calculation once in the constructor, make it a final instance variable ( class field ), and retrieve the same collection every time. Creating a new collection on every request can create an unnecessary Java memory overhead / make GarbageCollection work overtime.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        if(null==authGroups){
+            return Collections.emptySet();
+        }
+        Set<SimpleGrantedAuthority> grantedAuthorities = new HashSet<>();
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/security/ApplicationUserPrincipal.java]: it's better to add a 'final' modifier to the local variables to signal you don't intend re-assigning it later in the method block.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/service/UserService.java [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.module.User;
+import com.project.portfolio.springboot.onlinesellingandshopping.serverapp.repository.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+
+
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/service/UserService.java]: Needs Javadoc at the interface and methods level. See Effective Java Third Edition Item 56.

File: src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/service/UserServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+import javax.swing.text.html.Option;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class UserServiceImpl implements UserService{
+    @Autowired
+    UserRepository userRepository;
+    @Override
+    public Optional<User> getUser(String name){
```
Comment (natasha-gaysinsky) [inline src/main/java/com/project/portfolio/springboot/onlinesellingandshopping/serverapp/service/UserServiceImpl.java]: No null check. Effective Java Item 49.

File: src/main/resources/application.properties [hunk 15]
```diff
@@ -1 +1,14 @@
...
+spring.h2.console.enabled=true
+spring.datasource.url=jdbc:h2:mem:testdb;NON_KEYWORDS=user,VALUE
+spring.datasource.driverClassName=org.h2.Driver
+spring.datasource.username=sa
+spring.datasource.password=
+spring.jpa.hibernate.ddl-auto=update
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.H2Dialect
+
+###spring.jpa.properties.javax.persistence.schema-generation.database.action=drop-and-create
```
Comment (natasha-gaysinsky) [inline src/main/resources/application.properties]: It's best practice to remove commented out code / configuration. If you intend to use it, then maybe add TODO with a relevant comment to flag this?

File: src/main/resources/data.sql [hunk 16]
```diff
@@ -0,0 +1,7 @@
+
+
+
+INSERT INTO User VALUES(12,'danmargreen1','Danmar1','Green1','$2a$04$5j.ZgXSEc.v3w.s7yTbiRelZ73.AJfVd4l4Z.7soePVPLb6ZgTQ6O','5955 Skyline Dr, West Linn, OR, 97068','1996-01-17', 'danmargreen0@gmail.com','(503)-833-2453', NOW(), NOW());
+INSERT INTO User VALUES(14,'danmargreen2','Danmar2','Green2','$2a$10$SkyNcaKZ1ogQTMeSL9xIruwyqUyzsYd.pwR0twS/khEpS8hkzm6CC','5955 Skyline Dr, West Linn, OR, 97068','1996-01-17', 'danmargreen1@gmail.com','(503)-833-2453', NOW(), NOW());
+INSERT INTO Auth_User_Group VALUES(12,'danmargreen1', 'ROLE_ADMIN');
+INSERT INTO Auth_User_Group VALUES(14,'danmargreen2', 'ROLE_USER');
```
Comment (natasha-gaysinsky) [inline src/main/resources/data.sql]: Don't ask me why ( it has been bugging me as a developer :-) ) There needs to be a newline in the end of each file.

File: src/test/java/com/project/portfolio/springboot/onlinesellingandshopping/Test.java [hunk 17]
```diff
@@ -0,0 +1,13 @@
+package com.project.portfolio.springboot.onlinesellingandshopping;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Test {
```
Comment (natasha-gaysinsky) [inline src/test/java/com/project/portfolio/springboot/onlinesellingandshopping/Test.java]: This needs to be a unit test, where you mock dependencies for the class under test, and test happy path for each method, as well as corner cases. Take a look at the example here https://www.geeksforgeeks.org/how-to-write-test-cases-in-java-application-using-mockito-and-junit/

**Review by natasha-gaysinsky — APPROVED** [review]
Hi Danmar, congratulations on your first pull request for your Java project! I love the end to end thinking and approach you are taking to your first project. Awesome start!
